### ISSUE TITLE
CLI is stuck at "0% compiling" for some time

### ISSUE THREAD

**Issue by n9niwas** [angular/angular#36874]
# 🐞 Bug report

### Command

- [x] build
- [x] serve

### Is this a regression?

Yes, before enabling Ivy it didn't happen.

### Description

When running `ng serve` CLI is stuck at "0% compiling" for 1-1.5 minutes.
After some investigation I found that it happens because CLI runs `ngcc` processor.

We run `ngcc` in post-install script so it's just a waste of time.
Is it possible to completely disable ngcc in AngularCompilerPlugin?

Below is trace log:
<pre><code>
0% compiling
NgccProcessor.process: 3034.336ms
NgccProcessor.processModule.ngcc.process+tslib: 610.372ms
NgccProcessor.processModule.ngcc.process+@angular/core: 595.018ms
NgccProcessor.processModule.ngcc.process+@angularclass/hmr: 580.932ms
NgccProcessor.processModule.ngcc.process+rxjs: 561.337ms
NgccProcessor.processModule.ngcc.process+@angular/platform-server: 542.763ms
NgccProcessor.processModule.ngcc.process+@angular/platform-browser: 573.018ms
NgccProcessor.processModule.ngcc.process+@angular/common/http: 564.630ms
NgccProcessor.processModule.ngcc.process+@angular/animations/browser: 550.953ms
NgccProcessor.processModule.ngcc.process+@angular/platform-browser/animations: 558.415ms
NgccProcessor.processModule.ngcc.process+@angular/common: 566.523ms
NgccProcessor.processModule.ngcc.process+@angular/animations: 551.050ms
NgccProcessor.processModule.ngcc.process+@angular/router: 553.708ms
NgccProcessor.processModule.ngcc.process+@ngrx/store: 559.866ms
NgccProcessor.processModule.ngcc.process+@ngrx/effects: 543.515ms
NgccProcessor.processModule.ngcc.process+@nrwl/angular: 570.788ms
NgccProcessor.processModule.ngcc.process+@ngx-translate/core: 584.099ms
NgccProcessor.processModule.ngcc.process+rxjs/operators: 539.568ms
NgccProcessor.processModule.ngcc.process+he: 539.850ms
NgccProcessor.processModule.ngcc.process+hidden_private_module1: 550.323ms
NgccProcessor.processModule.ngcc.process+hidden_private_module2: 522.134ms
NgccProcessor.processModule.ngcc.process+hidden_private_module3: 595.114ms
NgccProcessor.processModule.ngcc.process+events: 531.826ms
NgccProcessor.processModule.ngcc.process+url: 547.360ms
NgccProcessor.processModule.ngcc.process+util: 538.700ms
NgccProcessor.processModule.ngcc.process+querystring: 536.670ms
NgccProcessor.processModule.ngcc.process+hidden_private_module4: 528.175ms
NgccProcessor.processModule.ngcc.process+qs: 551.256ms
NgccProcessor.processModule.ngcc.process+moment: 529.208ms
NgccProcessor.processModule.ngcc.process+@angular/forms: 515.195ms
NgccProcessor.processModule.ngcc.process+hidden_private_module5: 575.897ms
NgccProcessor.processModule.ngcc.process+@auth0/angular-jwt: 526.386ms
NgccProcessor.processModule.ngcc.process+oidc-client: 539.081ms
NgccProcessor.processModule.ngcc.process+ngx-cookie: 543.102ms
NgccProcessor.processModule.ngcc.process+apollo-angular: 544.677ms
NgccProcessor.processModule.ngcc.process+apollo-client: 561.253ms
NgccProcessor.processModule.ngcc.process+graphql-tag: 553.299ms
NgccProcessor.processModule.ngcc.process+apollo-link: 547.090ms
NgccProcessor.processModule.ngcc.process+graphql: 547.281ms
NgccProcessor.processModule.ngcc.process+zen-observable: 507.133ms
NgccProcessor.processModule.ngcc.process+zen-observable-ts: 545.172ms
NgccProcessor.processModule.ngcc.process+apollo-utilities: 545.245ms
NgccProcessor.processModule.ngcc.process+@wry/equality: 545.022ms
NgccProcessor.processModule.ngcc.process+apollo-cache: 541.682ms
NgccProcessor.processModule.ngcc.process+hidden_private_module6: 547.725ms
NgccProcessor.processModule.ngcc.process+apollo-angular-link-http: 540.515ms
NgccProcessor.processModule.ngcc.process+apollo-cache-inmemory: 556.511ms
NgccProcessor.processModule.ngcc.process+apollo-angular-link-http-common: 583.038ms
NgccProcessor.processModule.ngcc.process+optimism: 624.462ms
NgccProcessor.processModule.ngcc.process+@wry/context: 604.156ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/overlay: 604.266ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/scrolling: 633.058ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/portal: 628.590ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/bidi: 618.093ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/coercion: 609.324ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/collections: 598.324ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/platform: 603.611ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/a11y: 629.252ms
NgccProcessor.processModule.ngcc.process+@angular/cdk/observers: 617.754ms
NgccProcessor.processModule.ngcc.process+@angular/material/icon: 576.987ms
NgccProcessor.processModule.ngcc.process+@angular/material/core: 603.758ms
NgccProcessor.processModule.ngcc.process+resize-observer-polyfill: 572.928ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/dropdown: 603.869ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/modal: 580.077ms
NgccProcessor.processModule.ngcc.process+hidden_private_module7: 555.454ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/component-loader: 564.436ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/positioning: 560.246ms
NgccProcessor.processModule.ngcc.process+@angular/material/datepicker: 540.158ms
NgccProcessor.processModule.ngcc.process+@angular/material/button: 568.505ms
NgccProcessor.processModule.ngcc.process+@angular/material/dialog: 557.129ms
NgccProcessor.processModule.ngcc.process+@angular/material/form-field: 533.590ms
NgccProcessor.processModule.ngcc.process+@angular/material-moment-adapter: 547.598ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/popover: 543.341ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap: 559.584ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/chronos: 561.589ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/utils: 573.780ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/accordion: 554.829ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/alert: 605.471ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/buttons: 579.015ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/carousel: 596.644ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/collapse: 557.232ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/datepicker: 622.658ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/pagination: 581.819ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/progressbar: 568.361ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/rating: 531.200ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/sortable: 553.491ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/tabs: 559.255ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/timepicker: 536.328ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/tooltip: 534.024ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/typeahead: 541.616ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/mini-ngrx: 558.523ms
NgccProcessor.processModule.ngcc.process+ngx-bootstrap/locale: 590.425ms
NgccProcessor.processModule.ngcc.process+@ngrx/entity: 495.579ms
NgccProcessor.processModule.ngcc.process+@ngrx/router-store: 534.652ms
NgccProcessor.processModule.ngcc.process+@angular/platform-browser-dynamic: 519.414ms
NgccProcessor.processModule.ngcc.process+@angular/compiler: 540.801ms
NgccProcessor.processModule.ngcc.process+@ngrx/store-devtools: 539.909ms
NgccProcessor.processModule.ngcc.process+ngx-mask: 562.417ms
NgccProcessor.processModule.ngcc.process+@agm/core: 569.871ms
NgccProcessor.processModule.ngcc.process+fastclick: 564.724ms
NgccProcessor.processModule.ngcc.process+object-hash: 545.999ms
NgccProcessor.processModule.ngcc.process+google-maps: 542.602ms
NgccProcessor.processModule.ngcc.process+googlemaps: 548.793ms
NgccProcessor.processModule.ngcc.process+angular-svg-round-progressbar: 535.124ms
NgccProcessor.processModule.ngcc.process+uuid: 545.112ms
NgccProcessor.processModule.ngcc.process+date-fns: 530.173ms
NgccProcessor.processModule.ngcc.process+angular2-text-mask: 615.068ms
NgccProcessor.processModule.ngcc.process+file-saver: 558.931ms
NgccProcessor.processModule.ngcc.process+ng2-currency-mask: 558.394ms
NgccProcessor.processModule.ngcc.process+rxjs/webSocket: 579.907ms
NgccProcessor.processModule.ngcc.process+url-polyfill: 584.745ms
NgccProcessor.processModule.ngcc.process+jest-preset-angular: 624.755ms
AngularCompilerPlugin._createOrUpdateProgram.ng.createProgram: 73717.922ms
</code></pre>

## 🔬 Minimal Reproduction

Create new Angular application with CLI that has a lot of NPM dependencies.
Run `ng serve`.

## 🌍 Your Environment
<pre><code>
Angular CLI: 9.1.3
Node: 10.18.1
OS: darwin x64
Package manager: Yarn

Angular: 9.1.3
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.3
@angular-devkit/build-angular      0.901.3
@angular-devkit/build-ng-packagr   0.901.3
@angular-devkit/build-optimizer    0.901.3
@angular-devkit/build-webpack      0.901.3
@angular-devkit/core               9.1.3
@angular-devkit/schematics         9.0.3
@angular/cdk                       9.1.0
@angular/material                  9.1.0
@angular/material-moment-adapter   9.1.0
@ngtools/webpack                   9.1.3
@schematics/angular                9.1.3
@schematics/update                 0.901.3
ng-packagr                         9.1.0
rxjs                               6.5.4
typescript                         3.8.3
webpack                            4.42.0
</code></pre>

### Our dependencies
<pre><code>
{
  "scripts": {
    "ngcc": "ngcc",
    "postinstall": "yarn ngcc"
  },
  "dependencies": {
    "@agm/core": "1.1.0",
    "@angular/animations": "9.1.3",
    "@angular/cdk": "9.1.0",
    "@angular/common": "9.1.3",
    "@angular/compiler": "9.1.3",
    "@angular/core": "9.1.3",
    "@angular/forms": "9.1.3",
    "@angular/material": "9.1.0",
    "@angular/material-moment-adapter": "9.1.0",
    "@angular/platform-browser": "9.1.3",
    "@angular/platform-browser-dynamic": "9.1.3",
    "@angular/platform-server": "9.1.3",
    "@angular/router": "9.1.3",
    "@auth0/angular-jwt": "3.0.1",
    "@nestjs/common": "^6.8.3",
    "@nestjs/core": "^6.8.3",
    "@nestjs/jwt": "^6.1.1",
    "@nestjs/platform-express": "^6.8.3",
    "@nestjs/swagger": "^4.3.2",
    "@nestjs/typeorm": "^6.1.3",
    "@ngrx/effects": "9.0.0",
    "@ngrx/entity": "9.0.0",
    "@ngrx/router-store": "9.0.0",
    "@ngrx/store": "9.0.0",
    "@ngx-translate/core": "^11.0.0",
    "@ngx-translate/http-loader": "^4.0.0",
    "@nrwl/angular": "9.1.4",
    "angular-svg-round-progressbar": "^3.0.1",
    "angular2-text-mask": "^9.0.0",
    "apollo-angular": "^1.7.0",
    "apollo-angular-link-http": "^1.8.0",
    "apollo-cache-inmemory": "^1.6.3",
    "apollo-client": "^2.6.4",
    "apollo-link": "^1.2.13",
    "autoprefixer": "^9.6.0",
    "aws-sdk": "^2.622.0",
    "bootstrap": "^4.3.1",
    "bunyan": "^1.8.12",
    "cache-manager": "^2.10.0",
    "cache-manager-redis-store": "^1.5.0",
    "cheerio": "0.22.0",
    "class-transformer": "^0.2.3",
    "class-validator": "^0.11.0",
    "core-js": "^2.5.4",
    "date-fns": "^2.10.0",
    "dnscache": "^1.0.2",
    "domino": "2.1.3",
    "envalid": "^4.2.2",
    "fastclick": "^1.0.6",
    "file-saver": "^2.0.1",
    "font-awesome": "4.7.0",
    "google-maps": "^4.2.0",
    "graphite-udp": "^1.2.0",
    "graphql": "^14.5.8",
    "graphql-tag": "^2.10.1",
    "he": "^1.2.0",
    "helmet": "^3.21.1",
    "jscpd": "^1.2.3",
    "jsrsasign": "8.0.13",
    "libphonenumber-js": "^1.7.21",
    "lodash": "^4.17.12",
    "micromatch": "^3.1.9",
    "moment": "2.24.0",
    "monet": "^0.9.0",
    "mongodb": "^3.3.0",
    "ng2-currency-mask": "^5.3.1",
    "ngx-bootstrap": "4.3.0",
    "ngx-cookie": "4.0.2",
    "ngx-mask": "8.1.2",
    "object-hash": "1.3.0",
    "oidc-client": "^1.5.2",
    "openid-client": "^3.14.1",
    "qs": "^6.6.0",
    "redis": "^2.8.0",
    "reflect-metadata": "0.1.13",
    "resize-observer-polyfill": "1.5.1",
    "rxjs": "6.5.4",
    "rxjs-compat": "6.5.4",
    "sass-rem": "2.0.1",
    "swagger-ui-express": "^4.0.6",
    "tslib": "1.10.0",
    "typeorm": "^0.2.18",
    "uuid": "^3.3.2",
    "viz": "^0.0.1",
    "zone.js": "0.10.2"
  },
  "devDependencies": {
    "@angular-builders/custom-webpack": "9.1.0",
    "@angular-devkit/build-angular": "0.901.3",
    "@angular-devkit/build-ng-packagr": "0.901.3",
    "@angular-devkit/core": "9.1.3",
    "@angular/cli": "9.1.3",
    "@angular/compiler-cli": "9.1.3",
    "@angular/language-service": "9.1.3",
    "@angularclass/hmr": "2.1.3",
    "@commitlint/cli": "8.3.5",
    "@commitlint/config-conventional": "8.3.4",
    "@commitlint/lint": "8.3.5",
    "@commitlint/load": "8.3.5",
    "@nestjs/schematics": "^6.7.0",
    "@nestjs/testing": "^6.8.3",
    "@ngneat/spectator": "5.2.1",
    "@ngrx/schematics": "9.0.0",
    "@ngrx/store-devtools": "9.0.0",
    "@nrwl/cypress": "9.1.4",
    "@nrwl/jest": "9.1.4",
    "@nrwl/nest": "9.1.4",
    "@nrwl/node": "9.1.4",
    "@nrwl/workspace": "9.1.4",
    "@types/aws-lambda": "^8.10.43",
    "@types/bunyan": "^1.8.6",
    "@types/cheerio": "^0.22.10",
    "@types/cypress-axe": "^0.4.0",
    "@types/diff": "^4.0.2",
    "@types/ejs": "^2.6.0",
    "@types/fastclick": "^1.0.28",
    "@types/file-saver": "^2.0.0",
    "@types/googlemaps": "^3.38.1",
    "@types/graphite-udp": "^1.2.5",
    "@types/he": "^1.1.0",
    "@types/helmet": "^0.0.43",
    "@types/jasmine": "~2.8.6",
    "@types/jasminewd2": "~2.0.3",
    "@types/jest": "25.1.1",
    "@types/jsrsasign": "^8.0.2",
    "@types/lodash": "^4.14.116",
    "@types/lolex": "^3.1.1",
    "@types/mongodb": "^3.2.1",
    "@types/node-fetch": "^2.1.4",
    "@types/nunjucks": "^3.1.0",
    "@types/object-hash": "1.2.0",
    "@types/qs": "^6.5.1",
    "@types/recursive-readdir": "^2.2.0",
    "@types/redis": "^2.8.13",
    "@types/redis-mock": "^0.17.0",
    "@types/rimraf": "^2.0.2",
    "@types/uuid": "^3.4.4",
    "@types/webpack": "^4.41.6",
    "@types/xml2js": "^0.4.4",
    "@types/yargs": "^12.0.8",
    "angular-vscode": "^1.0.3",
    "babel-jest": "25.1.0",
    "babel-preset-jest": "^25.1.0",
    "chalk": "^2.4.2",
    "codelyzer": "5.1.2",
    "cypress": "^4.3.0",
    "cypress-axe": "^0.8.1",
    "diff": "^4.0.1",
    "dotenv": "6.2.0",
    "ejs": "^2.6.1",
    "glob": "^7.1.3",
    "husky": "^1.1.1",
    "ignore": "^5.1.4",
    "is-ci": "^1.1.0",
    "jasmine-marbles": "0.4.0",
    "jasmine-spec-reporter": "~4.2.1",
    "jest": "25.1.0",
    "jest-localstorage-mock": "2.4.0",
    "jest-preset-angular": "8.0.0",
    "lolex": "^3.0.0",
    "madge": "^3.2.0",
    "ng-mocks": "9.0.0",
    "ng-packagr": "9.1.0",
    "node-fetch": "^2.3.0",
    "npm-run-all": "^4.1.5",
    "nunjucks": "3.2.0",
    "prettier": "2.0.2",
    "pretty-quick": "2.0.1",
    "protractor": "~5.4.0",
    "recursive-readdir": "^2.2.2",
    "redis-mock": "^0.46.0",
    "rimraf": "^2.6.3",
    "rxjs-tslint-rules": "^4.10.0",
    "sass-extract": "2.1.0",
    "sass-extract-loader": "1.1.0",
    "stylelint": "9.3.0",
    "stylelint-config-recommended-scss": "3.2.0",
    "stylelint-scss": "3.1.3",
    "svgo": "^1.1.1",
    "tar": "^4.4.4",
    "targz": "1.0.1",
    "tmp": "^0.0.33",
    "ts-jest": "25.2.0",
    "ts-node": "~7.0.0",
    "tsconfig-paths": "^3.6.0",
    "tsickle": "^0.34.0",
    "tslint": "6.1.0",
    "tslint-angular": "3.0.2",
    "tslint-config-airbnb": "5.11.2",
    "tslint-config-prettier": "1.18.0",
    "tslint-sonarts": "1.9.0",
    "typescript": "3.8.3",
    "uncss": "^0.17.0",
    "url-polyfill": "^1.1.0",
    "webdriver-manager": "12.1.7",
    "webpack-bundle-analyzer": "^3.1.0",
    "xml2js": "^0.4.19",
    "yargs": "^12.0.5"
  }
}
</code></pre>

**Comment by mpk** [comment]
I have a similar problem with CLI being stuck at "0% compiling":

In my app, I have 324 npm modules that should be compiled by NGCC. First "ng serve" after npm install takes some time before all of them are compiled and cached, but that's fine. However, subsequent serves are also very slow (~ 9 minutes).

I have done some digging and found out that for each npm module (before mainNgcc bails out early due to cache), the following line is run (mainNgcc -> getSharedSetup -> readConfiguration):
[https://github.com/angular/angular/blob/master/packages/compiler-cli/src/perform_compile.ts#L197](https://github.com/angular/angular/blob/master/packages/compiler-cli/src/perform_compile.ts#L197)

This line parses and resolves application's tsconfig.json file (by walking through the entire project directory, which in my case is ~ 3700 files) and it takes about 1.1s to do so. However, since it is run for each npm module (in my case 324 modules), this results in 1.1s * 324 modules = **6 minutes** delay before the actual application start compiling.

I think the tsconfig.json should be parsed and resolved only once to avoid this delay.

**Comment by petebacondarwin** [comment]
The code that you are concerned about is in the Angular CLI @mpk. Please can you open an issue on that repo: https://github.com/angular/angular-cli

**Comment by mpk** [comment]
Maybe I don't fully understand how Angular code is split, but the referenced code is in angular/angular repository, not in angular/angular-cli, isn't it? (and this issue was recently transferred from angular-cli repo)

**Comment by petebacondarwin** [comment]
Oops!! Sorry. You are right. My mistake.

**Comment by petebacondarwin** [comment]
@mpk - yours is a different issue that should be dealt with in a different fix. So can we move it to a new issue, so that they can be tracked separately?

**Comment by petebacondarwin** [comment]
Here is an improvement that should help with this issue @n9niwas - https://github.com/angular/angular/pull/36881

**Comment by mpk** [comment]
@petebacondarwin OK, I have created a new issue - https://github.com/angular/angular/issues/36882

**Comment by petebacondarwin** [comment]
Thanks
